python-spec/src/somabase/_wrap.py

```diff
@@ -0,0 +1,220 @@
+        inst: base.Collection[_ST] = self
+        while isinstance(inst, CollectionProxy):
+            inst = inst._backing
+        return inst
+
+    # Sized
+
+    def __len__(self) -> int:
+        return len(self._backing)
+
+    # Iterable
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._backing)
+
+    # Container
+
+    def __contains__(self, key: Any) -> bool:
+        return key in self._backing
+
+    # Mapping
+
+    def __getitem__(self, key: str) -> _ST:
+        return self._backing[key]
+
+    @overload
+    def get(self, key: str) -> Optional[_ST]:
+        ...
+
+    @overload
+    def get(self, key: str, default: _T) -> Union[_ST, _T]:
+        ...
+
+    def get(
+        self, key: str, default: Union[_ST, _T, None] = None
+    ) -> Union[_ST, _T, None]:
+        return self._backing.get(key, default)
+
+    def keys(self) -> KeysView[str]:
+        return self._backing.keys()
+
+    def items(self) -> ItemsView[str, _ST]:
+        return self._backing.items()
+
+    def values(self) -> ValuesView[_ST]:
+        return self._backing.values()
+
+    # __reversed__ is not supported.
+
+    # MutableMapping
+
+    def __setitem__(self, key: str, value: _ST) -> None:
+        self._backing[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        del self._backing[key]
+
+    def pop(
+        self,
+        key: str,
+        value: Union[_T, _ST] = _SENTINEL,  # type: ignore[assignment]
+    ) -> Union[_T, _ST]:
+        if value is _SENTINEL:
+            return self._backing.pop(key)
+        return self._backing.pop(key, value)
+
+    def popitem(self) -> Tuple[str, _ST]:
+        return self._backing.popitem()
+
+    def clear(self) -> None:
+        self._backing.clear()
+
+    def update(self, __other=(), **kwds) -> None:
+        return self._backing.update(__other, **kwds)
+
+    def setdefault(
+        self,
+        key: str,
+        default: _ST = None,  # type: ignore[assignment]
+    ) -> _ST:
+        return self._backing.setdefault(key, default)  # type: ignore[arg-type]
+
+
+@attrs.define()
+class item(Generic[_T]):
+    """Descriptor to transform property access into indexing.
+
+    This descriptor works with :class:`CollectionProxy` to allow for simple
+    specification of properties that reflect the members of the collection::
+
+        class WrapImpl(CollectionProxy):
+
+            first = item(str)
+            second = item(int, "2nd")
+
+        inst = WrapImpl(some_collection)
+
+        # This is equivalent to getting some_collection["first"]
+        inst.first
+
+        # This is equivalent to setting some_collection["2nd"]
+        inst.second = 500
+    """
+
+    typ: Optional[Type[_T]] = None
+    """The type we expect to return from this field."""
+
+    item_name: Optional[str] = None
+    """The name of the item we are getting (``x._backing["whatever"]``).
+
+    This uses the name of the field by default but can be manually overridden.
```

Comment:
Is there a use case in mind to motivate having both `item_name` and `field_name`?  (I couldn't find any code that uses it, and am trying to understand the reason to have both)

Reply:
NM....I understand after reading the tests!